src/Qir/Runtime/lib/QSharpCore/intrinsics.cpp

```diff
@@ -37,46 +37,51 @@ static Microsoft::Quantum::IQuantumGateSet* GateSet()
 
 extern "C"
 {
-    void quantum__qis__exp__body(QirArray* paulis, double angle, QirArray* qubits)
+    void __quantum__qis__exp__body(QirArray* paulis, double angle, QirArray* qubits)
     {
         assert(paulis->count == qubits->count);
 
         std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
-        return GateSet()->Exp((long)(paulis->count), reinterpret_cast<PauliId*>(pauliIds.data()),
-                              reinterpret_cast<Qubit*>(qubits->buffer), angle);
+        GateSet()->Exp((long)(paulis->count), reinterpret_cast<PauliId*>(pauliIds.data()),
+                       reinterpret_cast<Qubit*>(qubits->buffer), angle);
     }
 
-    void quantum__qis__exp__adj(QirArray* paulis, double angle, QirArray* qubits)
+    void __quantum__qis__exp__adj(QirArray* paulis, double angle, QirArray* qubits)
     {
-        quantum__qis__exp__body(paulis, -angle, qubits);
+        __quantum__qis__exp__body(paulis, -angle, qubits);
     }
 
-    void quantum__qis__exp__ctl(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
+    void __quantum__qis__exp__ctl(QirArray* ctls, QirExpTuple* args)
     {
-        assert(paulis->count == qubits->count);
+        assert(args->paulis->count == args->targets->count);
 
-        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
-        return GateSet()->ControlledExp((long)(ctls->count), reinterpret_cast<Qubit*>(ctls->buffer),
-                                        (long)(paulis->count), reinterpret_cast<PauliId*>(pauliIds.data()),
-                                        reinterpret_cast<Qubit*>(qubits->buffer), angle);
+        std::vector<PauliId> pauliIds = ExtractPauliIds(args->paulis);
+        GateSet()->ControlledExp((long)(ctls->count), reinterpret_cast<Qubit*>(ctls->buffer),
+                                 (long)(args->paulis->count), reinterpret_cast<PauliId*>(pauliIds.data()),
+                                 reinterpret_cast<Qubit*>(args->targets->buffer), args->angle);
     }
 
-    void quantum__qis__exp__ctladj(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
+    void __quantum__qis__exp__ctladj(QirArray* ctls, QirExpTuple* args)
     {
-        quantum__qis__exp__ctl(ctls, paulis, -angle, qubits);
+        assert(args->paulis->count == args->targets->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(args->paulis);
+        GateSet()->ControlledExp((long)(ctls->count), reinterpret_cast<Qubit*>(ctls->buffer),
+                                 (long)(args->paulis->count), reinterpret_cast<PauliId*>(pauliIds.data()),
+                                 reinterpret_cast<Qubit*>(args->targets->buffer), -(args->angle));
     }
```

Comment:
Nit: Consider creating a local copy of `QirExpTuple`, negating the angle of it, and passing it to `quantum__qis__exp__ctl()`.
Otherwise currently `__quantum__qis__exp__ctladj()` nearly duplicates the whole `quantum__qis__exp__ctl()`.

Reply:
Ah, that's a good point. I had started considering that approach and then forgot to follow through after using this copy-paste just to verify behavior. I'll make that change so duplication is reduced.